core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java

```diff
@@ -176,14 +173,15 @@ private void completeRouter() {
   }
 
   private CoreEvent eventWithCurrentContextDeleted(CoreEvent event) {
-    Map<String, ForeachContext> foreachContextContainer = foreachContextResolver.getCurrentContextFromEvent(event);
-    foreachContextContainer.remove(event.getContext().getId());
-    return foreachContextResolver.eventWithContext(event, foreachContextContainer);
+    removeContext(event);
+    return quickCopy(event, new HashMap<>(0));
```

Comment:
what do you need this event copy for?

Reply:
This copy is needed to keep actual implementation. 
ForeachRouter creates an EventQuickCopyContextDecorator event instance to process items and returns an EventQuickCopyContextDecorator event instance when foreach ends.
quickCopy implementation is fixing a corner case when a huge number of java.util.Collections.unmodifiableMap are chained causing a StackOverFlow. 

https://github.com/mulesoft/mule/blob/887b71b617472914df85164373e93747ece8203d/core/src/main/java/org/mule/runtime/core/internal/event/DefaultEventBuilder.java#L592